 # Employee Attrition Analysis

## Exploring the Factors Behind Employee Turnover

This project analyzes employee data to identify the key factors associated with employee attrition and understand the characteristics of employees who leave the company.

The analysis covers employee demographics, job characteristics, satisfaction levels, income, overtime, business travel, and years of experience to uncover meaningful patterns in employee turnover.

 ## Project Objective

The main objective of this project is to explore employee attrition and identify the factors associated with employees leaving the company.

### Key Questions

- What is the overall employee attrition rate?
- Is overtime associated with higher attrition?
- How does job satisfaction relate to employee turnover?
- Which job roles have the highest attrition?
- Does income or age differ between employees who leave and those who stay?
- Does employee tenure affect attrition?
- Is business travel associated with higher employee turnover?

 ## Tools & Libraries

- **Python**
- **Pandas** — Data manipulation and analysis
- **NumPy** — Numerical operations
- **Plotly** — Interactive data visualization
- **Matplotlib** — Plot customization

## Dataset Overview

The dataset contains information about **1,470 employees** across **35 variables**.

The variables describe different aspects of employees, including:

- Demographics
- Job roles and departments
- Income and compensation
- Job satisfaction
- Overtime and business travel
- Work experience
- Career progression
- Employee attrition

The main target variable is **Attrition**, which indicates whether an employee left the company (`Yes`) or stayed (`No`).

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px

In [3]:
df = pd.read_csv("/kaggle/input/datasets/pavansubhasht/ibm-hr-analytics-attrition-dataset/WA_Fn-UseC_-HR-Employee-Attrition.csv")

In [4]:
df.head()

,Age,Attrition,BusinessTravel,DailyRate,Department,DistanceFromHome,Education,EducationField,EmployeeCount,EmployeeNumber,...,RelationshipSatisfaction,StandardHours,StockOptionLevel,TotalWorkingYears,TrainingTimesLastYear,WorkLifeBalance,YearsAtCompany,YearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager
0,41,Yes,Travel_Rarely,1102,Sales,1,2,Life Sciences,1,1,...,1,80,0,8,0,1,6,4,0,5
1,49,No,Travel_Frequently,279,Research & Development,8,1,Life Sciences,1,2,...,4,80,1,10,3,3,10,7,1,7
2,37,Yes,Travel_Rarely,1373,Research & Development,2,2,Other,1,4,...,2,80,0,7,3,3,0,0,0,0
3,33,No,Travel_Frequently,1392,Research & Development,3,4,Life Sciences,1,5,...,3,80,0,8,3,3,8,7,3,0
4,27,No,Travel_Rarely,591,Research & Development,2,1,Medical,1,7,...,4,80,1,6,3,3,2,2,2,2


In [5]:
df.shape

(1470, 35)

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 35 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Age                       1470 non-null   int64 
 1   Attrition                 1470 non-null   object
 2   BusinessTravel            1470 non-null   object
 3   DailyRate                 1470 non-null   int64 
 4   Department                1470 non-null   object
 5   DistanceFromHome          1470 non-null   int64 
 6   Education                 1470 non-null   int64 
 7   EducationField            1470 non-null   object
 8   EmployeeCount             1470 non-null   int64 
 9   EmployeeNumber            1470 non-null   int64 
 10  EnvironmentSatisfaction   1470 non-null   int64 
 11  Gender                    1470 non-null   object
 12  HourlyRate                1470 non-null   int64 
 13  JobInvolvement            1470 non-null   int64 
 14  JobLevel                

In [7]:
df.describe().T

,count,mean,std,min,25%,50%,75%,max
Age,1470.0,36.923810,9.135373,18.0,30.00,36.0,43.00,60.0
DailyRate,1470.0,802.485714,403.509100,102.0,465.00,802.0,1157.00,1499.0
DistanceFromHome,1470.0,9.192517,8.106864,1.0,2.00,7.0,14.00,29.0
Education,1470.0,2.912925,1.024165,1.0,2.00,3.0,4.00,5.0
EmployeeCount,1470.0,1.000000,0.000000,1.0,1.00,1.0,1.00,1.0
EmployeeNumber,1470.0,1024.865306,602.024335,1.0,491.25,1020.5,1555.75,2068.0
EnvironmentSatisfaction,1470.0,2.721769,1.093082,1.0,2.00,3.0,4.00,4.0
HourlyRate,1470.0,65.891156,20.329428,30.0,48.00,66.0,83.75,100.0
JobInvolvement,1470.0,2.729932,0.711561,1.0,2.00,3.0,3.00,4.0
JobLevel,1470.0,2.063946,1.106940,1.0,1.00,2.0,3.00,5.0


In [8]:
print(df.isnull().sum().sum())


0


In [9]:
print(df.duplicated().sum())

0


In [10]:
categorical_cols = df.select_dtypes(include="object").columns

for col in categorical_cols:
    print(f"\n")
    print(df[col].value_counts())



Attrition
No     1233
Yes     237
Name: count, dtype: int64


BusinessTravel
Travel_Rarely        1043
Travel_Frequently     277
Non-Travel            150
Name: count, dtype: int64


Department
Research & Development    961
Sales                     446
Human Resources            63
Name: count, dtype: int64


EducationField
Life Sciences       606
Medical             464
Marketing           159
Technical Degree    132
Other                82
Human Resources      27
Name: count, dtype: int64


Gender
Male      882
Female    588
Name: count, dtype: int64


JobRole
Sales Executive              326
Research Scientist           292
Laboratory Technician        259
Manufacturing Director       145
Healthcare Representative    131
Manager                      102
Sales Representative          83
Research Director             80
Human Resources               52
Name: count, dtype: int64


MaritalStatus
Married     673
Single      470
Divorced    327
Name: count, dtype: int64


Over18
Y    1

In [11]:
df.nunique().sort_values()

EmployeeCount                  1
Over18                         1
StandardHours                  1
Attrition                      2
OverTime                       2
PerformanceRating              2
Gender                         2
BusinessTravel                 3
Department                     3
MaritalStatus                  3
RelationshipSatisfaction       4
StockOptionLevel               4
JobSatisfaction                4
EnvironmentSatisfaction        4
JobInvolvement                 4
WorkLifeBalance                4
Education                      5
JobLevel                       5
EducationField                 6
TrainingTimesLastYear          7
JobRole                        9
NumCompaniesWorked            10
PercentSalaryHike             15
YearsSinceLastPromotion       16
YearsWithCurrManager          18
YearsInCurrentRole            19
DistanceFromHome              29
YearsAtCompany                37
TotalWorkingYears             40
Age                           43
HourlyRate

In [12]:
df = df.drop(columns=[
    "EmployeeCount",
    "StandardHours",
    "Over18",
    "EmployeeNumber"
])

In [13]:
df.shape

(1470, 31)

In [14]:
print(df["Attrition"].value_counts())

print('-' * 40)

print(df["Attrition"].value_counts(normalize=True) * 100)

Attrition
No     1233
Yes     237
Name: count, dtype: int64
----------------------------------------
Attrition
No     83.877551
Yes    16.122449
Name: proportion, dtype: float64


## 3. Exploratory Data Analysis

In this section, we explore the main factors associated with employee attrition using interactive visualizations and descriptive analysis.

### Employee Attrition Overview

We first examine the overall distribution of employee attrition to understand how common employee turnover is in the dataset.

In [15]:
attrition_counts = df["Attrition"].value_counts().reset_index()
attrition_counts.columns = ["Attrition", "Count"]

fig = px.bar(
    attrition_counts,
    x = "Attrition",
    y = "Count",
    title = "Employee Attrition"
)

fig.show()

# ****The majority of employees stayed with the company, while a smaller proportion left, indicating that employee attrition is relatively limited compared to employee retention.****

In [16]:
overtime_attrition = (df.groupby("OverTime")["Attrition"].value_counts(normalize = True).mul(100).rename("Percentage").reset_index())

overtime_attrition = overtime_attrition[overtime_attrition["Attrition"] == "Yes"]

overtime_attrition

,OverTime,Attrition,Percentage
1,No,Yes,10.436433
3,Yes,Yes,30.528846


### Attrition Rate by Overtime

We examine whether working overtime is associated with a higher employee attrition rate.

In [17]:
fig = px.bar(
    overtime_attrition,
    x = "OverTime",
    y = "Percentage",
    text = "Percentage",
    title = "Attrition Rate by Overtime"
)

fig.update_traces(
    texttemplate = "%{text:.1f}%",
    textposition = "outside"
)

fig.update_layout(
    yaxis_title = "Attrition Rate (%)",
    xaxis_title = "OverTime"
)

fig.show()

# ****Employees who work overtime have a noticeably higher attrition rate than employees who do not, suggesting that overtime may be an important factor associated with employee turnover.****

In [18]:
satisfaction_attrition = (df.groupby("JobSatisfaction")["Attrition"].value_counts(normalize = True).mul(100).rename("Percentage").reset_index())

satisfaction_attrition = satisfaction_attrition[satisfaction_attrition["Attrition"] == "Yes"]

satisfaction_attrition

,JobSatisfaction,Attrition,Percentage
1,1,Yes,22.837370
3,2,Yes,16.428571
5,3,Yes,16.515837
7,4,Yes,11.328976


### Attrition Rate by Job Satisfaction

We analyze whether employee job satisfaction is associated with employee turnover.

In [19]:
fig = px.bar(
    satisfaction_attrition,
    x = "JobSatisfaction",
    y = "Percentage",
    text = "Percentage",
    title = "Attrition Rate by Job Satisfaction"
)

fig.update_traces(
    texttemplate = "%{text:.1f}%",
    textposition = "outside"
)

fig.update_layout(
    xaxis_title = "Job Satisfaction Level",
    yaxis_title = "Attrition Rate (%)"
)

fig.show()

# Insight: Employees with the lowest job satisfaction (Level 1) have the highest attrition rate at 22.8%, while employees with the highest satisfaction (Level 4) have the lowest attrition rate at 11.3%. This suggests that lower job satisfaction is associated with a higher likelihood of employee turnover.# 

In [20]:
role_attrition = (df.groupby("JobRole")["Attrition"].value_counts(normalize = True).mul(100).rename("Percentage").reset_index())

role_attrition = role_attrition[role_attrition["Attrition"] == "Yes"].sort_values("Percentage", ascending = False)

role_attrition

,JobRole,Attrition,Percentage
17,Sales Representative,Yes,39.759036
5,Laboratory Technician,Yes,23.938224
3,Human Resources,Yes,23.076923
15,Sales Executive,Yes,17.484663
13,Research Scientist,Yes,16.095890
9,Manufacturing Director,Yes,6.896552
1,Healthcare Representative,Yes,6.870229
7,Manager,Yes,4.901961
11,Research Director,Yes,2.500000


### Attrition Rate by Job Role

We compare attrition rates across different job roles to identify roles with higher employee turnover.

In [21]:
fig = px.bar(
    role_attrition,
    x = "Percentage",
    y = "JobRole",
    orientation="h",
    text = "Percentage",
    title = "Attrition Rate by Job Role"
)

fig.update_traces(
    texttemplate = "%{text:.1f}%",
    textposition = "outside"
)

fig.update_layout(
    xaxis_title = "Attrition Rate (%)",
    yaxis_title = "Job Role"
)

fig.show()

# Insight: Attrition varies significantly across job roles. Sales Representatives have the highest attrition rate at 39.8%, followed by Laboratory Technicians (23.9%) and Human Resources (23.1%). In contrast, Research Directors have the lowest attrition rate at only 2.5%. This suggests that employee turnover is strongly associated with job role.

### Monthly Income and Attrition

We compare the distribution of monthly income between employees who left the company and those who stayed.

In [22]:
fig = px.box(
    df,
    x = "Attrition",
    y = "MonthlyIncome",
    title = "Monthly Income Distribution by Attrition",
    points = False
)

fig.update_layout(
    xaxis_title = "Attrition",
    yaxis_title = "Monthly Income"
)

fig.show()

# Insight: Employees who left the company generally had lower monthly incomes than those who stayed. The median monthly income for employees who left is around $3.2K, compared with approximately $5.2K for employees who stayed, suggesting that lower income is associated with higher employee attrition.

### Age and Attrition

We compare employee age distributions between employees who left and those who remained with the company.

In [23]:
fig = px.box(
    df,
    x = "Attrition",
    y = "Age",
    title = "Age Distribution by Attrition",
    points = False
)

fig.update_layout(
    xaxis_title = "Attrition",
    yaxis_title = "Age"
)

fig.show()

# Insight: Employees who left the company tend to be younger, with a median age of approximately 32 years, compared with around 36 years for employees who stayed. This suggests that younger employees are more likely to leave the company.

In [24]:
bins = [0, 2, 5, 10, 20, 50]
labels = ["0-2", "3-5", "6-10", "11-20", "21+"]

df["YearsAtCompanyGroup"] = pd.cut(
    df["YearsAtCompany"],
    bins = bins,
    labels = labels,
    include_lowest = True
)

In [25]:
years_attrition = (df.groupby("YearsAtCompanyGroup", observed = False)["Attrition"].value_counts(normalize = True).mul(100).rename("Percentage").reset_index())

years_attrition = years_attrition[years_attrition["Attrition"] == "Yes"]

### Attrition Rate by Years at Company

We examine whether employee tenure is associated with different levels of attrition.

In [26]:
fig = px.bar(
    years_attrition,
    x = "YearsAtCompanyGroup",
    y = "Percentage",
    text ="Percentage",
    title = "Attrition Rate by Years at Company"
)

fig.update_traces(
    texttemplate = "%{text:.1f}%",
    textposition = "outside"
)

fig.update_layout(
    xaxis_title = "Years at Company",
    yaxis_title = "Attrition Rate (%)"
)

fig.show()

# Insight: Employees who have been with the company for 0–2 years have the highest attrition rate at 29.8%, while employees with 11–20 years have the lowest rate at 6.7%. This suggests that employees are most likely to leave during their early years at the company.

In [27]:
travel_attrition = (df.groupby("BusinessTravel")["Attrition"].value_counts(normalize = True).mul(100).rename("Percentage").reset_index())

travel_attrition = travel_attrition[travel_attrition["Attrition"] == "Yes"]

### Attrition Rate by Business Travel

We investigate whether the frequency of business travel is associated with employee attrition.

In [28]:
fig = px.bar(
    travel_attrition,
    x = "BusinessTravel",
    y = "Percentage",
    text = "Percentage",
    title = "Attrition Rate by Business Travel"
)

fig.update_traces(
    texttemplate = "%{text:.1f}%",
    textposition = "outside"
)

fig.update_layout(
    xaxis_title = "Business Travel",
    yaxis_title = "Attrition Rate (%)"
)

fig.show()

# Insight: Employees who travel frequently for business have the highest attrition rate at 24.9%, compared with 15.0% for employees who travel rarely and only 8.0% for non-traveling employees. This suggests that frequent business travel is associated with a higher likelihood of employee turnover.

In [29]:
df_corr = df.copy()

df_corr["Attrition"] = df_corr["Attrition"].map({"Yes": 1, "No": 0})

corr = df_corr.select_dtypes(include="number").corr()

### Correlation Analysis

We examine correlations between numerical variables to identify relationships that may be associated with employee attrition.

In [30]:
fig = px.imshow(
    corr,
    text_auto = ".2f",
    aspect = "auto",
    title = "Correlation Heatmap"
)

fig.update_layout(
    width = 1000,
    height = 900
)

fig.show()

# Insight: Attrition shows weak negative correlations with several employee characteristics, including TotalWorkingYears (-0.17), JobLevel (-0.17), Age (-0.16), and MonthlyIncome (-0.16). This indicates that employees with greater experience, higher job levels, older ages, and higher incomes tend to have slightly lower attrition. However, the correlations are relatively weak, suggesting that no single numerical factor strongly explains employee turnover.

In [31]:
role_overtime_attrition = (df.groupby(["JobRole", "OverTime"])["Attrition"].apply(lambda x: (x == "Yes").mean() * 100).reset_index(name = "AttritionRate"))

### Attrition Rate by Job Role and Overtime

We combine job role and overtime to investigate whether the relationship between overtime and attrition varies across different job roles.

In [32]:
fig = px.bar(
    role_overtime_attrition,
    x = "JobRole",
    y = "AttritionRate",
    color = "OverTime",
    barmode = "group",
    text = "AttritionRate",
    title = "Attrition Rate by Job Role and Overtime"
)

fig.update_traces(
    texttemplate = "%{text:.1f}%",
    textposition = "outside"
)

fig.update_layout(
    xaxis_title = "Job Role",
    yaxis_title = "Attrition Rate (%)",
    xaxis_tickangle = -45
)

fig.show()

# Insight: Overtime is associated with substantially higher attrition in several job roles. The strongest differences appear among Sales Representatives (66.7% vs. 28.8%), Laboratory Technicians (50.0% vs. 15.7%), and Human Resources (38.5% vs. 17.9%) for employees working overtime versus those who do not. However, this pattern is not consistent across every role, indicating that the relationship between overtime and attrition may vary by job role.

## Conclusion

The analysis reveals several important patterns associated with employee attrition.

- Overtime is associated with higher employee attrition.
- Lower job satisfaction is associated with higher turnover.
- Sales Representatives have the highest overall attrition rate among job roles.
- Employees in their first 2 years at the company show the highest attrition rate.
- Frequent business travelers have higher attrition compared with non-travelers.
- Employees who leave tend to be younger and have lower monthly incomes.
- The combination of job role and overtime highlights particularly high attrition in some roles.

Overall, employee attrition appears to be influenced by a combination of job satisfaction, workload, tenure, compensation, business travel, and job role rather than a single factor.

These findings can help HR teams identify groups with higher turnover risk and investigate potential improvements in employee retention.

# ## Key Insights

- Employees who work overtime show a higher attrition rate than those who do not.
- Employees with low job satisfaction have higher attrition rates, with the highest rate of 22.8% at satisfaction level 1.
- Sales Representatives have the highest overall attrition rate among job roles at 39.8%.
- Employees with 0–2 years at the company have the highest attrition rate at 29.8%, indicating that early-tenure employees are more likely to leave.
- Frequent business travelers have a higher attrition rate (24.9%) compared with non-travelers (8.0%).
- Employees who left the company generally had lower monthly incomes and were younger than those who stayed.
- Correlation analysis shows that Age, JobLevel, MonthlyIncome, and TotalWorkingYears have weak negative relationships with Attrition.
- The combination of Job Role and Overtime reveals particularly high attrition among overtime employees in some roles, especially Sales Representatives and Laboratory Technicians.

In [34]:
df = df.drop(
    columns = ["YearsAtCompanyGroup", "DistanceGroup"],
    errors = "ignore"
)

In [35]:
print("Shape:", df.shape)

print("\nMissing values:", df.isnull().sum().sum())

print("Duplicate rows:", df.duplicated().sum())

print("\nData types:")
print(df.dtypes.value_counts())

Shape: (1470, 31)

Missing values: 0
Duplicate rows: 0

Data types:
int64     23
object     8
Name: count, dtype: int64


In [36]:
df.head()

,Age,Attrition,BusinessTravel,DailyRate,Department,DistanceFromHome,Education,EducationField,EnvironmentSatisfaction,Gender,...,PerformanceRating,RelationshipSatisfaction,StockOptionLevel,TotalWorkingYears,TrainingTimesLastYear,WorkLifeBalance,YearsAtCompany,YearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager
0,41,Yes,Travel_Rarely,1102,Sales,1,2,Life Sciences,2,Female,...,3,1,0,8,0,1,6,4,0,5
1,49,No,Travel_Frequently,279,Research & Development,8,1,Life Sciences,3,Male,...,4,4,1,10,3,3,10,7,1,7
2,37,Yes,Travel_Rarely,1373,Research & Development,2,2,Other,4,Male,...,3,2,0,7,3,3,0,0,0,0
3,33,No,Travel_Frequently,1392,Research & Development,3,4,Life Sciences,4,Female,...,3,3,0,8,3,3,8,7,3,0
4,27,No,Travel_Rarely,591,Research & Development,2,1,Medical,1,Male,...,3,4,1,6,3,3,2,2,2,2


## Key Takeaway

Employee attrition is not driven by one single factor. The analysis suggests that workload, job satisfaction, career stage, compensation, business travel, and job role all play a role in employee turnover.

Understanding these patterns can help organizations focus their employee retention efforts on the areas with the highest potential impact.